source/server/admin/prometheus_stats_request.cc

```diff
@@ -0,0 +1,160 @@
+#include "source/server/admin/prometheus_stats_request.h"
+
+#include <string>
+#include <vector>
+
+#include "source/server/admin/prometheus_stats.h"
+
+#include "stats_params.h"
+#include "stats_render.h"
+
+namespace Envoy {
+namespace Server {
+
+namespace {
+bool metricLessThan(const Stats::RefcountPtr<Stats::Metric>& stat1,
+                    const Stats::RefcountPtr<Stats::Metric>& stat2) {
+  ASSERT(&stat1->constSymbolTable() == &stat2->constSymbolTable());
+  return stat1->constSymbolTable().lessThan(stat1->statName(), stat2->statName());
+}
+} // namespace
+
+PrometheusStatsRequest::PrometheusStatsRequest(Stats::Store& stats, const StatsParams& params,
+                                               Stats::CustomStatNamespaces& custom_namespaces,
+                                               UrlHandlerFn url_handler_fn)
+    : StatsRequestBase(stats, params, url_handler_fn), custom_namespaces_(custom_namespaces) {
+
+  // the "type" query param is ignored for prometheus stats, so always start from
+  // counters; also, skip the TextReadouts phase unless that stat type is explicitly
+  // requested via query param
+  if (params_.prometheus_text_readouts_) {
+    phases_ = {Phase{PhaseName::Counters, "Counters"}, Phase{PhaseName::Gauges, "Gauges"},
+               Phase{PhaseName::TextReadouts, "Text Readouts"},
+               Phase{PhaseName::Histograms, "Histograms"}};
+  } else {
+    phases_ = {Phase{PhaseName::Counters, "Counters"}, Phase{PhaseName::Gauges, "Gauges"},
+               Phase{PhaseName::Histograms, "Histograms"}};
+  }
+  phase_ = 0;
+}
+
+template <class StatType> Stats::IterateFn<StatType> PrometheusStatsRequest::checkStat() {
+  return [this](const Stats::RefcountPtr<StatType>& stat) -> bool {
+    // check if unused
+    if (params_.used_only_ && !stat->used()) {
+      return true;
+    }
+
+    // check if filtered
+    if (params_.filter_ != nullptr) {
+      if (!std::regex_search(stat->name(), *params_.filter_)) {
+        return true;
+      }
+    } else if (params_.re2_filter_ != nullptr &&
+               !re2::RE2::PartialMatch(stat->name(), *params_.re2_filter_)) {
+      return true;
+    }
+
+    // capture stat by either adding to a pre-existing variant or by creating a new variant, and
+    // then storing the variant into the map
+    const Stats::SymbolTable& global_symbol_table = stat->constSymbolTable();
+    std::string tag_extracted_name = global_symbol_table.toString(stat->tagExtractedStatName());
+
+    StatOrScopes variant = stat_map_.contains(tag_extracted_name)
```

Comment:
you can make this a little faster (just one hash lookup rather than 3 in some cases) with:

```
StatOrScopes& variant = stat_map_[tag_extracted_name];
if (variant.index() == absl::variant_npos) {
  variant = std::vector<Stats::RefcountPtr<StatType>>({stat});
} else {
  absl::get<std::vector<Stats::RefcountPtr<StatType>>>(variant).emplace_back(stat);
}
```